src/main/java/uk/gov/companieshouse/api/accounts/interceptor/TransactionInterceptor.java

```diff
@@ -28,26 +28,63 @@
     @Override
     @SuppressWarnings("unchecked")
     public boolean preHandle(HttpServletRequest request, HttpServletResponse response,
-            Object handler) {
+        Object handler) {
         try {
             Map<String, String> pathVariables = (Map) request
-                    .getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
+                .getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
             String transactionId = pathVariables.get("transactionId");
             ResponseEntity<Transaction> transaction = transactionManager
-                    .getTransaction(transactionId, request.getHeader("X-Request-Id"));
+                .getTransaction(transactionId, request.getHeader("X-Request-Id"));
             request.setAttribute(AttributeName.TRANSACTION.getValue(), transaction.getBody());
-            return isTransactionIsOpen(transaction);
+
+            return isValidTransaction(request, transaction);
+
         } catch (HttpClientErrorException httpClientErrorException) {
             response.setStatus(httpClientErrorException.getStatusCode().value());
             return false;
         }
     }
 
+    /**
+     * Check transaction is valid. e.g. Accounts api call needs the Transaction to be open. Filing
+     * Generator call needs the transaction to be closed.
+     *
+     * @param request
+     * @param transaction
+     * @return
+     */
+    private boolean isValidTransaction(HttpServletRequest request,
+        ResponseEntity<Transaction> transaction) {
+        if (isFilingGeneratorRequest(request)) {
```

Comment:
This method could quickly get out of hand as we add the need to check different transaction statuses for different reasons. Is it worth refactoring this filing logic out into a different interceptor?

Initially, it sounds like the drawback would be that we'd arguably need to get the transaction status again, meaning another potential call to the Transactions API - however, we could add logic in both interceptors (this and the new one) that checks the request to see if it has the transaction added to it already (see line #38 that sets the transaction on the request). If it doesn't have the transaction info, it goes and gets it from the Transactions API, otherwise it just gets the details from the request. 

Alternatively, we could add order precedence when adding the interceptors in `CompanyAccountsApplication` and guarantee that the interceptor that retrieves the transaction is called first (see https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/servlet/config/annotation/InterceptorRegistration.html#order-int-), however I think this approach is flawed as it makes assumed knowledge about a different interceptor, tightly coupling them and making them difficult to refactor.

We could also factor out the `isTransactionIsOpen`/`isTransactionClosed` methods into some kind of utility, to help promote re-usability of code.

It feels like this interceptor is too general purpose, and runs the risk of becoming unwieldy. Any kind of separation of responsibility we can promote would benefit us strongly in the long run.

Reply:
I could agree more with this comment. 
> This method could quickly get out of hand as we add the need to check different transaction statuses for different reasons.

I can see that interceptor was designed to cater for the accounts api calls, but it seems Filing Generator was never taken  into consideration when designing it. It might be worth to check the interceptor's design again to future proof it. 

@drosser, this may be conversation we should have to discuss with @sroberts2 @gpatrac.

Reply:
Regarding this, 
> Initially, it sounds like the drawback would be that we'd arguably need to get the transaction status again, meaning another potential call to the Transactions API 

I don't see your pont in here: `meaning another potential call to the Transactions API`.
The transaction's information is already retrieved before calling the method, and there will not be any extra calls to the transaction api: 
```JAVA
ResponseEntity<Transaction> transaction = transactionManager
    .getTransaction(transactionId, request.getHeader("X-Request-Id"));
    .getTransaction(transactionId, request.getHeader("X-Request-Id"));
request.setAttribute(AttributeName.TRANSACTION.getValue(), transaction.getBody());
```

Reply:
Regarding this: 
> however, we could add logic in both interceptors (this and the new one) that checks the request to see if it has the transaction added to it already (see line #38 that sets the transaction on the request). If it doesn't have the transaction info, it goes and gets it from the Transactions API, otherwise it just gets the details from the request.

This is more design issue, and it has been decided to get the transaction at the fist entry point, `TransactionInterceptor`.  We wouldn't need get the `transaction` information in the other interceptors

Reply:
Regarding this: 
>Alternatively, we could add order precedence when adding the interceptors in CompanyAccountsApplication and guarantee that the interceptor that retrieves the transaction is called first (see https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/servlet/config/annotation/InterceptorRegistration.html#order-int-), however I think this approach is flawed as it makes assumed knowledge about a different interceptor, tightly coupling them and making them difficult to refactor.

> We could also factor out the isTransactionIsOpen/isTransactionClosed methods into some kind of utility, to help promote re-usability of code.

> It feels like this interceptor is too general purpose, and runs the risk of becoming unwieldy. Any kind of separation of responsibility we can promote would benefit us strongly in the long run.

This is more design issue, @drosser, this may be conversation we should have to discuss with @sroberts2 @gpatrac.